### PR TITLE
Feature/save and load analysis

### PR DESCRIPTION
**Pull request by alxvth**
This PR supersedes #455.
All existing plugins should still just compile and run fine with after PR.

ManiVault can now save and load analysis plugins to/from projects files.

Analysis plugins now can have multiple Input and Output datasets:
```cpp
// Analysis & Transformation and Writer Plugins
void setInputDataset(const Dataset<DatasetImpl>& inputDataset);
void setInputDatasets(const Datasets& inputDatasets);               // NEW

Dataset<DatasetType> getInputDataset();
Datasets getInputDatasets();                                        // NEW

// Only Analysis Plugins
void setOutputDataset(const Dataset<DatasetImpl>& outputDataset);
void setOutputDatasets(const Datasets& outputDatasets);             // NEW

Dataset<DatasetType> getOutputDataset();
Datasets getOutputDatasets();                                       // NEW

bool outputDataInit()                                               // NEW
```

Your Analysis plugin must of course overload the `fromVariantMap` and `toVariantMap` functions. Additionally, the `init()` function needs to take into account that the out dataset of the plugin will already be loaded and does not need to be created anymore.
Here is an example setup when implementing save-able analysis plugins:
```cpp
// if not loaded from project, init new data
if (!outputDataInit())
{
    qDebug() << "PCAPlugin::create new derived data";

    auto newOutput = Dataset<Points>(mv::data().createDerivedDataset("PCA", inputDataset, inputDataset));
    setOutputDataset(newOutput);

    // Set initial data (default 2 dimensions, all points at (0,0) )
    std::vector<float> initialData;
    const size_t numInitialDataDimensions = 2;
    const auto numPoints = inputDataset->getNumPoints();
    initialData.resize(numInitialDataDimensions * numPoints);
    newOutput->setData(initialData.data(), numPoints, numInitialDataDimensions);
    events().notifyDatasetDataChanged(newOutput);
}

// Same as always
auto outputDataset = getOutputDataset<Points>();
```

Other changes:
- Serialization of `DatasetImpl::_properties`
- Serialization of `DatasetImpl::_groupIndex`
- Serialization of `DatasetImpl::_mayUnderive`
- Added `PointData::getVersion` which returns the current ManiVault version. This helps with adjusting behavior for loading projects which saved point data whose version was "No Version"
- Fix serialization of `PointData::_dimNames`: They only correctly saved when `getDimensionNames().size() != getNumPoints()`
- Marked more functions as const

### CODE REVIEW

**Comment by alxvth** [comment]
`Plugin::getVersion()` returns now the version string set in the meta data in the `plugin.json` file

**Comment by alxvth** [comment]
This PR also addresses #200 with 
```cpp
    if (variantMap.contains("GroupIndex") && variantMap["GroupIndex"].toInt() >= 0)
        setGroupIndex(variantMap["GroupIndex"].toInt());
```
See [here](https://github.com/ManiVaultStudio/core/pull/457/files#diff-d247288e68d47c768da7460b26a995560fde016f594363160e8fa67de0de66baR165)

**Review by ThomasKroes — CHANGES_REQUESTED** [review]
I only have five small comments (see in the PR conversation) :)

I will test this PR in concert with the [SNE PR](https://github.com/ManiVaultStudio/t-SNE-Analysis/pull/88) ASAP!

File: HDPS/src/Set.cpp [hunk 1]
```diff
@@ -194,10 +212,13 @@ QVariantMap DatasetImpl::toVariantMap() const
         { "PluginKind", QVariant::fromValue(_rawData->getKind()) },
         { "PluginVersion", QVariant::fromValue(_rawData->getVersion()) },
         { "Derived", QVariant::fromValue(isDerivedData()) },
+        { "MayUnderive", QVariant::fromValue(mayUnderive()) },
+        { "Full", QVariant::fromValue(isFull()) },
+        { "SourceDatasetGUID", isDerivedData() ? QVariant::fromValue(_sourceDataset->getId()) : "" },
```
[the remark this answers is not in the document]
Comment (alxvth) [reply HDPS/src/Set.cpp]: I changed all GUIDs to IDs

**Pull request merged**